src/wallet/wallet.cpp

```diff
@@ -70,6 +71,47 @@ const CWalletTx* CWallet::GetWalletTx(const uint256& hash) const
     return &(it->second);
 }
 
+// Generate a new spending key and return its public payment address
+CZCPaymentAddress CWallet::GenerateNewZKey()
+{
+    AssertLockHeld(cs_wallet); // mapZKeyMetadata
+    auto k = SpendingKey::random();
+    auto addr = k.address();
+
+    // Check for collision, even though it is unlikely to ever occur
+    if (CCryptoKeyStore::HaveSpendingKey(addr))
+        throw std::runtime_error("CWallet::GenerateNewZKey(): Collision detected");
+
+    // Create new metadata
+    int64_t nCreationTime = GetTime();
+    mapZKeyMetadata[addr] = CKeyMetadata(nCreationTime);
+
+    CZCPaymentAddress pubaddr(addr);
+    if (!AddZKey(k))
+        throw std::runtime_error("CWallet::GenerateNewZKey(): AddZKey failed");
+    return pubaddr;
+}
+
+// Add spending key to keystore and persist to disk
+bool CWallet::AddZKey(const libzcash::SpendingKey &key)
+{
+    AssertLockHeld(cs_wallet); // mapZKeyMetadata
+    auto addr = key.address();
+
+    if (!CCryptoKeyStore::AddSpendingKey(key))
+        return false;
+
+    if (!fFileBacked)
+        return true;
+
+    if (!IsCrypted()) {
+        return CWalletDB(strWalletFile).WriteZKey(addr,
+                                                  key,
+                                                  mapZKeyMetadata[addr]);
+    }
```

Comment:
Upstream behaviour is that when a new taddr is generated or imported, the key is added to the wallet and saved to disk immediately. It can be tricky to see this from the code due to polymorphic behaviour, but here is what happens.  Bear in mind that CWallet inherits from CCryptoKeyStore.
```
pwalletMain->AddKeyPubKey(key, pubkey)
--> CWallet::AddKeyPubKey
  --> CCryptoKeyStore::AddKeyPubKey
    --> AddCryptedKey(pubkey, vchCryptedSecret)
     --> Since no class specified, polymorphic behaviour should be to call method based on object class (which is a wallet and not the base class which just adds the keys to the internal map)
     --> CWallet::AddCryptedKey
       calls both in sequence:
       --> CCryptoKeyStore::AddCryptedKey (update map in memory)
       --> WriteCryptedKey (update db on disk)
```